chai/cmd_executor.py

```diff
@@ -0,0 +1,225 @@
+
+# TODO remove `type: ignore` when stubs are available for grpc.aio See
+# https://github.com/shabbyrobe/grpc-stubs/issues/22
+import grpc.aio as aio  # type: ignore
+from typing_extensions import Self
+
+import chai.client as client
+import chai.cmdExecutor_pb2 as msg
+import chai.cmdExecutor_pb2_grpc as service
+from chai.client import RpcResult
+
+Input = client.Source.Input
+# Derived from JSON encodings
+Counterexample = dict
+TlaModule = dict
+
+
+class UnexpectedErrorException(Exception):
+    """For unexpected application errors"""
+
+
+@dataclass
+class CmdExecutorError(ABC):
+    """Base class for known application errors
+
+    Attributes:
+        pass_name: The name of the processing pass that produced the error.
+    """
+
+    pass_name: str
+
+    # Derive an error from decoded JSON
+    @abstractclassmethod
+    def _of_dict(cls, _: dict) -> Self:
+        ...
+
+
+def _ensure_is_pass_failure(d: dict) -> None:
+    if d["error_type"] != "pass_failure":
+        raise UnexpectedErrorException(f"Unexpected error receieved from RPC call: {d}")
+
+
+@dataclass
+class ParsingError(CmdExecutorError):
+    """Records a parsing error
+
+    Attributes:
+        errors: A list of parsing error messages.
+    """
+
+    errors: List[str]
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError:
+        _ensure_is_pass_failure(d)
+
+        data = d["data"]
+        pass_name = data["pass_name"]
+        if pass_name == "SanyParser":
+            return ParsingError(pass_name, data)
+        else:
+            raise UnexpectedErrorException(
+                f"Unexpected error receieved from RPC call: {d}"
+            )
+
+
+@dataclass
+class TypecheckingError(CmdExecutorError):
+    """Records a typechecking error
+
+    Attributes:
+        errors: A list of tuples pairing source locations with type error
+            messages.
+    """
+
+    errors: List[Tuple[str, str]]  # location, msg errors
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError | TypecheckingError:
+        _ensure_is_pass_failure(d)
+
+        pass_name = d["data"]["pass_name"]
+        errors = d["data"]["error_data"]
+        if pass_name == "TypeCheckerSnowcat":
+            return TypecheckingError(pass_name, errors)
+        else:
+            return ParsingError._of_dict(d)
+
+
+@dataclass
+class CheckingError(CmdExecutorError):
+    """Records a model checking error
+
+    Attributes:
+        checking_result: The kind of model checking result. The possible result
+            kinds are as follows
+
+            - Error: A checking violation is found.
+            - Deadlock: A deadlock was found.
+            - RuntimeError: A runtime error was encountered, preventing checking.
+        counter_examples: A list of counterexamples found.
+    """
+
+    checking_result: str
+    counter_example: List[Counterexample]
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError | TypecheckingError | CheckingError:
```

Comment:
In terms of structure, I find it a bit weird that a classmethod on `*Error` also produces other kinds of errors. Indicates to me that all this code for constructing errors really belongs in its own logical unit. WDYT?

Reply:
I see what you mean. I think I am confusing two things here: there are `CheckingError`s which indicate an error resulting from the checking pass, and then there are the errors that can be returned from the checking command RPC, which belong to this union.

Differentiating the errors for a phase from the set of errors that can be produced from an RPC makes sense. Thanks for that clarification :+1:

Reply:
Do you find 279dd4142b68a9685d33e7d478ab3987bcf57969 improves the situation here?